pysd/py_backend/external.py

```diff
@@ -467,6 +396,70 @@ def _initialize_data(self, dim_name):
 
         return data
 
+    def _fill_missing(self, series, data):
+        """
+        Fills missing values in excel read data. Mutates the values in data.
```

Comment:
Can you explain here how the fill happens, and how the user can control the behavior? Alternately, if there are other places where the user decides how to handle missing data, do we have appropriate docstrings there to make sure that they know how to do it, and what to expect?

Reply:
When Vensim founds a missing value, it leaves it missing and then when it needs to calculate its value uses the interpolation method defined for the object. This can be seen in the output of a model with DATA variable with missing values.
Missing will add another time column with the times where data values are available or missing, at the right we will have the data values. For example:

|Time     | Value1 | Value2 |
| --------- |:---------:| ----------:|
| 0 | | 1 |
|1| 2 | |
|2| 4 | 5 |
| 3| 6 | |
|4|||

Then Vensim during simulation will use Value 1 as (1, 2), (2, 4), (3, 6) and Value2 as (0, 1),  (2, 5) and use the interpolation method defined for the data (interpolate, hold backward, lock forward or raw). Vensim will return different types of warnings:

```
WARNING: Simulation starts before first time for exogenous variable - Value1.
WARNING: Exogenous variable - Value2 - in range at time 1.
WARNING: Simulating beyond range for exogenous variable - Value2 - at time 3.
WARNING: Simulating beyond range for exogenous variable - Value1 - at time 4.
```

So Vensim does not fill the values but use shorter series without the missing values, then why filling missing values with PySD, the reason is simplicity. With the upper example let's suppose that we have a subscripted variable

|Time     | Var[A] | Var[B] |
| --------- |:---------:| ----------:|
| 0 | | 1 |
|1| 2 | |
|2| 4 | 5 |
| 3| 6 | |
|4|||

Then we will generate an xarray with the given values, which will have nan values where missing values are. This will crash when we try to use the implemented routines for hold backward and look forward. We will need to add a check each time step to see where we have missing values and which value return, and do that for each coordinate, which may create a lot of overhead. In the case of the xarray interp built-in method it will crash as we have nan values in the dataarray.
Then filling this values during the initialization will make the code mutch simple and faster, but to keep it work like Vensim, we need to fill them with the interpolation method used during integration, that is why `self.interp` attribute is used.

So, in order to maintain the certainty that everything is working well, I will add a warning by default as Vensim (in the future we can add an option to ignore or raise an error, but is not my priority now). The cases of GET DATA with interp are tested with `test_get_with_missing_values_xlsx`, but I will complete it for regular GET LOOKUPS and GET DATA with look forward and keep backward.

I think it will be interesting to add a user-friendly way to interact with missing values and change their value. In any case, this should be studied because we need to define a good way to interact with external objects. I may work on that option in the future as I will need to add user interaction to change the name of Excel files and sheets that are read.

Reply:
I have just noticed that Vensim does something very weird with missing values for lookups, it just raises a warning but them the output variable is filled with 0. So for this particular case, I will no add a test in the test-models, because it is clearly a Vensim's bug, then I will then raise an error

Reply:
Hi @JamesPHoughton . I have already made several changes. I am waiting for you to accept the PR in the test-models repo to add the new commit to PySD. Then I will push the changes :)

Reply:
Had a chat with @tseyanglim who knows vensim's "features" intimately. Here's what he says:
> There is a setting in the import that lets you select whether to fill with a default or not. Said setting doesn't play well between Vensim 7 and Vensim 8. So e.g. a control file (.frm) with import settings for, if I recall, Vensim 8 that says not to fill the values, if used with a Vensim 7 import, will fill them anyway.

Raising the error in PySD is a good idea.

Reply:
The setting in question: